### PR TITLE
Untranslated text discrepancy to be highlighted in blue

### PR DESCRIPTION
**Pull request by dungeonmaster51**
https://dimagi-dev.atlassian.net/browse/ICDS-1709

SUMMARY
This PR makes improvements to Commcare translation checker script

Checks for missing translations or English words in regional languages and highlights the text discrepancy in blue

### CODE REVIEW

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 1]
```diff
@@ -140,6 +148,54 @@ def convertCellToOutputValueList(cell):
...
+    Input:
+    cell (xl.cell.cell.Cell): Cell whose contents are to be parsed
+
+    Output:
+    Dict with strings as key and occurence as value
+
+   """
+    outputDict = {}
+    x = cell.value
+    if'jr://file/' in x:
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:164]: space after if

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 2]
```diff
@@ -140,6 +148,54 @@ def convertCellToOutputValueList(cell):
...
+    try:
+        x = re.sub(r'(?:\s)<output[^, ]*', '', x)
+        x = re.sub(r'(?:\s)value=[^, ]*', '', x)
+    except Exception as e:
+        raise FatalError("FATAL ERROR determining string values for worksheet %s cell %s : %s" %
+                         (cell.parent.title, cell.coordinate, str(e)))
+    
+    for i in x.split():
+        i = re.sub(r'[\W\_]','',i)
+        if not outputDict.get(i) and (i and i.strip()) and not (re.match(r'^[0-9]*$', i)):
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:176]: If a word occurs more than once, do we increment occurrence count or we ignore it.

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 3]
```diff
@@ -140,6 +148,54 @@ def convertCellToOutputValueList(cell):
...
+
+    return outputDict
+
+def linguisticCharChecker(baseDict, colDict):
+    """
+    takes base column word dictionary and current column word dictionary as input. 
+    Returns list of common words and whether the script is english in both the columns or not. 
+
+   """
+    sharedWords = list(set(baseDict).intersection(colDict))
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:187]: Do we just compare only the shared words or even the occurrence count. For example, if base text has word w 2 times, and translated text has same word w only once, what do we do?

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 4]
```diff
@@ -439,7 +510,8 @@ def checkRowForMismatch(row, columnDict, fixedColumnDict, baseColumnIdx=None, ig
         mismatchCell.value = "Y"
         mismatchCell.style = curMismatchFillStyle
     else:
-        mismatchCell.value = "N"
+        if(mismatchCell.value!="Y"):
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:513]: bracket not need. Also please use pep-8 styling. There's no spaces after commas and general formatting issues.

**Review by adil-uddin — COMMENTED** [review]
Some questions and general formatting required.

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 1 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py]: updated.

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 4 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py]: updated.

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 2 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py:176]: We take only unique elements in the dictionary. The keys are unique and value will always be 1. We do not increment count if the word occurs again.

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 3 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py:187]: We take only unique words in the dictionary and value will always be 1 in the dictionary. Hence this situation doesn't arise.

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 5]
```diff
@@ -140,6 +148,54 @@ def convertCellToOutputValueList(cell):
...
+
+    try:
+        x = re.sub(r'(?:\s)<output[^, ]*', '', x)
+        x = re.sub(r'(?:\s)value=[^, ]*', '', x)
+    except Exception as e:
+        raise FatalError("FATAL ERROR determining string values for worksheet %s cell %s : %s" %
+                         (cell.parent.title, cell.coordinate, str(e)))
+    
+    for i in x.split():
+        i = re.sub(r'[\W\_]','',i)
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:175]: Nit: spaces after commas

File: CommcareTranslationChecker/CommcareTranslationChecker.py [hunk 6]
```diff
@@ -439,7 +510,8 @@ def checkRowForMismatch(row, columnDict, fixedColumnDict, baseColumnIdx=None, ig
         mismatchCell.value = "Y"
         mismatchCell.style = curMismatchFillStyle
     else:
-        mismatchCell.value = "N"
+        if mismatchCell.value!="Y":
```
Comment (adil-uddin) [inline CommcareTranslationChecker/CommcareTranslationChecker.py:513]: Nit: space around = operator

**Review by adil-uddin — APPROVED** [review]
Looks good to me, except for minor nits. @mkangia

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 5 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py]: Updated

File: CommcareTranslationChecker/CommcareTranslationChecker.py (hunk 6 above)
Comment (dungeonmaster51) [reply CommcareTranslationChecker/CommcareTranslationChecker.py]: updated

**Comment by mkangia** [comment]
@adil-uddin all resolved on this now?

Do we need to test this once before merging?

**Comment by adil-uddin** [comment]
yes @mkangia I think this is good to go but better to test it once. If @dungeonmaster51 can post the results of his testing here that'd be great

**Comment by mkangia** [comment]
@adil-uddin I don't think @dungeonmaster51 would be around to test this. Can you do a test? I hope it will be quick and won't take long. let me know if you are not able to find time for this.

**Comment by adil-uddin** [comment]
image_2020_09_28T13_49_58_714Z
@mkangia @dungeonmaster51 I just tested with exact scenario mentioned in the ticket and it doesn't seem to detect the missing translation ?

**Comment by mkangia** [comment]
Closing in favor of https://github.com/dimagi/CommcareTranslationChecker/pull/28

**Pull request closed without merging**